src/common/hooks/useLoginGuard/index.tsx

```diff
@@ -0,0 +1,51 @@
+import dynamic from 'next/dynamic'
+
+import { useState } from 'react'
+
+import { UniversalDialogProps } from '@/common/components/UniversalDialog/types'
+import { userStore } from '@/store/userStore'
+
+const DynamicUniversalDialog = dynamic(
+  async () =>
+    (
+      await import(
+        /* webpackChunkName: "UniversalDialog" */
+        '@/common/components/UniversalDialog'
+      )
+    ).UniversalDialog,
+  {
+    ssr: false,
+  }
+)
+
+export const useLoginGuard = () => {
+  const [open, setOpen] = useState(false)
+  const [onConfirm, setOnConfirm] = useState<(() => void) | undefined>(() => () => {})
+
+  // TODO better way to handle this?
+  const Dialog = () => (
+    <DynamicUniversalDialog
+      open={open}
+      setOpen={setOpen}
+      heading="คุณยังไม่ได้เข้าสู่ระบบ"
+      content="คุณจะสามารถเขียนรีวิว กดไลก์ และรายงานรีวิวของผู้อื่นได้เมื่อเข้าสู่ระบบแล้วเท่านั้น"
+      primaryButtonText="เข้าสู่ระบบ"
+      secondaryButtonText="ยกเลิก"
+      onConfirm={() => {
+        onConfirm?.()
+        userStore.login()
+      }}
+    />
+  )
```

Comment:
```suggestion
  const dialog = (
    <DynamicUniversalDialog
      open={open}
      setOpen={setOpen}
      heading="คุณยังไม่ได้เข้าสู่ระบบ"
      content="คุณจะสามารถเขียนรีวิว กดไลก์ และรายงานรีวิวของผู้อื่นได้เมื่อเข้าสู่ระบบแล้วเท่านั้น"
      primaryButtonText="เข้าสู่ระบบ"
      secondaryButtonText="ยกเลิก"
      onConfirm={() => {
        onConfirm?.()
        userStore.login()
      }}
    />
  )
```

And consumers can do `{dialog}` instead of `<Dialog />`

Reply:
shouldn't do that way. It WORKS but it's about convention. `dialog` may be just a function, not a component. To ensure it's a component we should named it with Capital letter.

Reply:
If you care about convention that much, you wouldn't define a component in a functional component.